salt/modules/xbps-pkg.py

```diff
@@ -0,0 +1,641 @@
+import os
+import re
+import logging
+
+# Import salt libs
+import salt.utils
+import salt.utils.decorators as decorators
+from salt.exceptions import CommandExecutionError, MinionError
+
+# Import 3rd-party libs
+import salt.ext.six as six
+
+log = logging.getLogger(__name__)
+
+# Define the module's virtual name
+__virtualname__ = 'pkg'
+
+
+@decorators.memoize
+def _check_xbps():
+    '''
+    Looks to see if xbps-install is present on the system, return full path
+    '''
+    return salt.utils.which('xbps-install')
+
+
+@decorators.memoize
+def _get_version():
+    '''
+    Get the xbps version
+    '''
+    xpath = _check_xbps()
+    version_string = __salt__['cmd.run'](
+        '{0} --version'.format(xpath), output_loglevel='trace'
+    )
+    if version_string is None:
+        # Dunno why it would, but...
+        return False
+
+    VERSION_MATCH = re.compile(r'(?:XBPS:[\s]+)([\d.]+)(?:[\s]+.*)')
+    version_match = VERSION_MATCH.search(version_string)
+    if not version_match:
+        return False
+
+    return version_match.group(1).split('.')
+
+
+def _rehash():
+    '''
+    Recomputes internal hash table for the PATH variable.
+    Used whenever a new command is created during the current
+    session.
+    '''
+    shell = __salt__['environ.get']('SHELL')
+    if shell.split('/')[-1] in ('csh', 'tcsh'):
+        __salt__['cmd.run']('rehash', output_loglevel='trace')
+
+
+def __virtual__():
+    '''
+    Set the virtual pkg module if the os is Void and xbps-install found
+    '''
+    if __grains__['os'] in ( 'Void') and _check_xbps():
+        return __virtualname__
+    return False
+
+
+def list_pkgs(versions_as_list=False, **kwargs):
+    '''
+    List the packages currently installed as a dict::
+
+        {'<package_name>': '<version>'}
+
+    .. versionadded:: XXX 201X.XX
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' pkg.list_pkgs
+    '''
+    versions_as_list = salt.utils.is_true(versions_as_list)
+    # not yet implemented or not applicable
+    if any([salt.utils.is_true(kwargs.get(x))
+            for x in ('removed', 'purge_desired')]):
+        return {}
+
+    cmd = 'xbps-query -l'
+    ret = {}
+    out = __salt__['cmd.run'](cmd, output_loglevel='trace')
+    for line in out.splitlines():
+        if not line:
+            continue
+        try:
+            # xbps-query -l output sample: 
+            # ii desktop-file-utils-0.22_4  Utilities to ...
+            #
+            # XXX handle package status (like 'ii') ?
+            pkg, ver = line.split(None)[1].rsplit('-', 1)
+        except ValueError:
+            log.error('xbps-query: Unexpected formatting in '
+                      'line: "{0}"'.format(line))
+
+        __salt__['pkg_resource.add_pkg'](ret, pkg, ver)
+    
+    __salt__['pkg_resource.sort_pkglist'](ret)
+    if not versions_as_list:
+        __salt__['pkg_resource.stringify'](ret)
+    return ret
+
+
+def list_upgrades(**kwargs):
```

Comment:
This would be better if we could take explicit arguments, as the other package providers typically do for this function.